### PR TITLE
Challenge "sort numbers"-strategies list: sort/ pop up/ insert/Math.min()

### LINKED ISSUE #6: Challenge "split object": key/value
Create and edit challenge "split object" file:

- [x] sandbox.js
- [x] split-object.spec.js

### LINKED ISSUE #12: Challenge "sort numbers" strategies list: sort/ pop up/ insert/Math.min()
Create and edit challenge "sort numbers" file:

- [x] sandbox.js
- [x] sort-numbers.spec.js

### PR DESCRIPTION
**Pull request by AlinaTaoRao**
---
name: solution code review
about: A template PR for code review, complete with checklist
title: 'Code Review: challenge name, strategy name'
---
close #12  

Before the reviewer merges this PR to main/master, all of these boxes must be
checked. Only the reviewer can check boxes.

## Behavior

> What the function does - inputs and outputs

- **The Function**
  - [x] The function's name is clear and helpful
  - [x] The parameter names are clear and helpful
  - [x] Default parameters are used if possible
- **JSDoc Comment**
  - [x] The docstring is well-formatted
  - [x] There are no spelling mistakes
  - [x] The description of the function's behavior is clear
  - [x] All parameters are documented with types
  - [x] Documented parameter names match the function's parameter names
  - [x] Parameters have a good description
  - [x] If default parameters are used, they are the correct type
  - [x] The return value is documented with a type
  - [x] The return value has a clear description
- **README**
  - [ ] All the sections in the solution README are complete
  - [ ] Descriptions in the README are thorough and clear
  - [ ] There is at least 2 use-cases
- **Unit Tests**
  - [x] Unit tests are present
  - [ ] There are at least 20 test cases
  - [x] All `describe`s and `it`s have clear and helpful names
  - [x] All test names actually describe what is being tested
  - [x] Tests are grouped logically using `describe`
  - [x] There is only one `expect` inside every `it`
  - [x] There is no logic (loops, conditionals, ...) in the test cases
  - [x] Side-effects are tested, if necessary
  - [x] All tests pass in the CI check
  - [x] The tests have 100% coverage of the solution
  - [x] Tricky edge cases are tested (extra credit)

## Strategy

> The logic behind the solution

- **The Function**
  - [x] Variables have names that help understand the strategy
  - [x] Spacing and formatting help to understand the strategy
  - [x] The code is written as simply and clearly as possible
  - [x] Use short, clear comments to describe the strategy
- **README**
  - [ ] There is an explanation of the solution's strategy in the README
  - [ ] The explanation is clear and understandable
  - [ ] The explanation actually matches the function that was written
## Implementation

> The code used to write the solution

- **Formatting**
  - [x] No formatting errors appear in the linting check
- **Linting**
  - [x] There should be no linting errors or warnings in the CI check
  - [x] Spelling should be correct in all code, comments and docstrings
- **The solution file**
  - [ ] Has a JSDoc comment
  - [ ] Has an exported function
  - [ ] Does no other code or comments outside of the function body (you can use
        `sandbox.js` for saving extra ideas)
- **The function body _Should_ ...**
  - [ ] Be well and consistently formatted
  - [ ] Have clear and helpful variable names
  - [ ] Declare variables in the smallest scope possible
  - [ ] Be a simple and readable implementation
- **The function body should _Not_ ...**
  - [ ] Have extra comments that are not about the solution's strategy
  - [ ] Have any user interactions (`prompt`, `alert`, `confirm`)
  - [ ] Have any `console` calls (`.log`, `.assert`, ...)
  - [ ] Have any `debugger` statements

### CODE REVIEW

**Comment by AlinaTaoRao** [comment]
@aviv82 @sannesofie 

Hi, I've added `3rd try` . My branch is ready to merge to main. Please check it. Thanks in advance. 😄

File: src/sort-numbers/alina/sandbox.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ */
+/* 
+
+strategy:
+1. build a Compare function.
+2. use .sort() to return an array with ascending as default sort order.
+*/
+
+// 1st Try: inspired by Evan. //works
+export const sortNumbers1 = (number = []) => {
```
Comment (MichielPovre) [inline src/sort-numbers/alina/sandbox.js]: argument name is a bit doubtful -> at first glance I assumed the argument was just one number value, but its an arrayOfNumbers. it will read a bit more smooth if you clarify this in the name of the argument

File: src/sort-numbers/alina/sort-numbers.spec.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import { sortNumbers1, sortNumbers2, sortNumbers3 } from './sandbox.js';
+
+for (const sortNumbers of [sortNumbers1, sortNumbers2, sortNumbers3]) {
+  describe('sortNumbers sorts an array of numbers', () => {
```
Comment (MichielPovre) [inline src/sort-numbers/alina/sort-numbers.spec.js]: `describe('sortNumbers sorts an array of numbers'` -> i won't be able to link the test suite to each different implementation (sortNumbers1,2 & 3) like this. (the report will say "sortNumbers sorts an array of numbers" for all 3 implementations, but I want it to say "sortNumbersX sorts an array of numbers" for implementation X. suggestion: `describe(sortNumbers.name + " sorts an array of numbers"...`

File: src/sort-numbers/alina/sort-numbers.spec.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import { sortNumbers1, sortNumbers2, sortNumbers3 } from './sandbox.js';
```
Comment (MichielPovre) [inline src/sort-numbers/alina/sort-numbers.spec.js]: great test cases! If i would only look at the tests & not the implementation itself, I would already be confident your function behaves as i would expect.

File: src/sort-numbers/alina/sandbox.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  const copyOfNumber = number.map((x) => x);
+  return copyOfNumber.sort((firstEl, secondEl) => firstEl - secondEl);
+};
+
+// 2nd Try: use pop up. // works
+export const sortNumbers2 = (number = []) => {
+  const copyOfNumber = number.map((x) => x);
+  for (let j = 0; j <= copyOfNumber.length - 2; j++) {
+    let indexOfMin = j;
+    for (let i = j + 1; i < copyOfNumber.length; i++) {
```
Comment (MichielPovre) [inline src/sort-numbers/alina/sandbox.js]: `    for (let i = j + 1; i < copyOfNumber.length; i++) {
      if (copyOfNumber[indexOfMin] > copyOfNumber[i]) {
        indexOfMin = i;
      }` -> this code looks for the index of the minimum. to avoid the need of this extra forloop, you could take a look at the indexOf function, in combination with a math function  (Math.min(...copyOfNumber) returns the minimum value. combine these two in : `copyOfNumber.indexOf(Math.min(...copyOfNumber))` and this oneliner could replace this for-loop :)

File: src/sort-numbers/alina/sandbox.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    const temp = copyOfNumber[j];
+    copyOfNumber[j] = copyOfNumber[indexOfMin];
+    copyOfNumber[indexOfMin] = temp;
+  }
+  return copyOfNumber;
+};
+
+// 3rd Try: use insert. // works
+export const sortNumbers3 = (number = []) => {
```
Comment (MichielPovre) [inline src/sort-numbers/alina/sandbox.js]: just a comment: nice to see other concepts (break) & use for loops in different ways. However, keep in mind that this implementation does 3 for-loops and its quite hard to see the strategy here -> maybe a small comment above each for loop that states what the loop is for, could already make it easier to understand for reviewers :)

**Review by MichielPovre — COMMENTED** [review]
great work :) three good and different implementations + a test suite that covers the behaviour of the challenge totally. Some small remarks & comments left in the review

File: src/sort-numbers/alina/sandbox.js (hunk 1 above)
Comment (AlinaTaoRao) [reply src/sort-numbers/alina/sandbox.js]: Use arrayOfNumbers instead of number as argument name. Checked.

File: src/sort-numbers/alina/sort-numbers.spec.js (hunk 3 above)
Comment (AlinaTaoRao) [reply src/sort-numbers/alina/sort-numbers.spec.js]: I copied test unit from @colevandersWands. He wrote a good test and I just did a little change to make it suitable for testing multipole functions. 😄

File: src/sort-numbers/alina/sandbox.js (hunk 4 above)
Comment (AlinaTaoRao) [reply src/sort-numbers/alina/sandbox.js]: Very useful advice.  👍 I will use these methods to build a 4th try.

File: src/sort-numbers/alina/sandbox.js (hunk 5 above)
Comment (AlinaTaoRao) [reply src/sort-numbers/alina/sandbox.js]: working on it.

File: src/sort-numbers/alina/sort-numbers.spec.js (hunk 2 above)
Comment (AlinaTaoRao) [reply src/sort-numbers/alina/sort-numbers.spec.js]: > describe(sortNumbers.name + " sorts an array of numbers"...

very good advice. 👍 Working on it.

**Comment by AlinaTaoRao** [comment]
@MichielPovre 
Thanks for giving me so many constructive advice. :angel: I learned a lot. :eyes: :writing_hand:  :muscle: :raised_hands:

File: src/sort-numbers/alina/sandbox.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+strategy:
+1. build a Compare function.
+2. use .sort() to return an array with ascending as default sort order.
+*/
+
+// 1st Try: inspired by Evan. //works
+export const sortNumbers1 = (arrayOfNumbers = []) => {
+  const copy = arrayOfNumbers.map((x) => x);
+  return copy.sort((firstEl, secondEl) => firstEl - secondEl);
```
Comment (aviv82) [inline src/sort-numbers/alina/sandbox.js:28]: cool. i wonder what this is

**Review by aviv82 — APPROVED** [review]
very nicely done!
